Brubeck Client for Browser

timoxley opened the pull request:
Adds support for browser builds of client.



* Added some additional steps + verification to browser tests. Now uses two clients with different keys and adds grant permissions + assign storage steps.
* Fixed CORS requests for requests coming from client to storage node.
* Removed rootDirs from typescript config, was confusing and used inconsistently anyway.
Now you have to `import … from ../../src/file` instead of `import … from ../../file`
* *Made protocol a direct dependency of broker.*

#### Improved bundling process and reduced size

* Split Logger, Tracker & Network Node so browser build can import node without any tracker dependencies e.g. express
* Replaced ethers & bip39 (mnemonic) packages in protocol with equivalent individual `ethersproject/*` packages. This reduces bundle size.
* Removed double bundling of core-js, just use usage plugin from `@babel/preset-env (See https://github.com/zloirock/core-js/issues/848#issuecomment-717208058)
* Updated dev dependencies & unified dependency versions across packages.

#### Improved bootstrap time: 1m30s to 12s 
* Removed unnecessary build steps e.g. clean before build
* Reusable tsc builds using `tsc --build` mode. https://www.typescriptlang.org/docs/handbook/project-references.html
* Added caching to eslint

#### Also Note
* ~Some diff noise due to reverting some merges from main~ Reverted those reverts.

### New Updates
* Integrated npm workspaces https://github.com/streamr-dev/network-monorepo/pull/259
* Fixed all eslint warnings.
* Fixed network browser/karma tests.
* Removed unused client dependencies.
* Updated client dependencies.
* Added a fix for resend flakiness.
* Prevented dist, build, node_modules, etc from being copied into docker image, it installs rebuilds things anyway.
* Set up client integration tests & cross client tests to run against monorepo brokers/trackers.
* Configured husky to run eslint before pushing
* Enabled client export tests in CI.
* Removed lerna usage everywhere but the show-versions script.

timoxley committed:
test(client): Add more config merging tests.

timoxley committed:
build(network): Fix error during karma webpack step: 'Multiple assets emit different content to the same name'.

timoxley committed:
fix(broker): Fix CORS requests when client connects directly to storage node.

timoxley committed:
build(network): Build network package in production mode before npm publishing.

timoxley committed:
build(network): Split up node, tracker, composition & Logger code for more precision about what goes into browser builds.

timoxley committed:
fix(client): Make browser builds functional.

timoxley committed:
ci(client): Re-enable browser build & test workflow.

timoxley committed:
test(client): Mark late subscriber tests as resend tests.

timoxley committed:
test(broker): Cleaner NodeMetrics test teardown.

timoxley committed:
test(client): Debug browser tests.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  Broker: remove eslint exceptions (#245)
  test(client): fix validation test
  protocol: function return types (#242)

timoxley committed:
deps(protocol): Replace ethers & bip39 packages with individual ethersproject libs.

timoxley committed:
build(cli-tools, network): Cache eslint.

timoxley committed:
deps(client): Update dev deps. Unify monorepo versions & update package-locks.

timoxley committed:
build(client): Reduce browser bundle size.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  ci(broker): rm extranous test-sequential run (#246)

timoxley committed:
test(client): Check error handling of Pipeline pull.

timoxley committed:
fix(client): Allow subscriptions using onMessage to bubble errors.

timoxley committed:
test(client): Tidy tests.

timoxley committed:
fix(client): Fail resends early if stream fetch fails.

timoxley committed:
build: Improve monorepo bootstrap build time with tsc --build mode. Broker now depends on protocol directly.

timoxley committed:
build: Fix up minor tsconfig issues.

timoxley committed:
deps: Refresh package-locks.

timoxley committed:
build(client): Fix client tsconfig include.

timoxley committed:
build(client, network): Browser streams/resends working correctly.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  fix(network): Node: rm unused constructor opts
  network: unit test for TrackerConnector (#250)
  network [NET-264]: rework node propagation logic (#211)
  broker: remove Todo type (#248)
  Update plugins.md
  Update README.md
  Update README.md
  DEVOP-102: run broker Docker build in self hosted runner (#249)
  network: refactor inbound/outboundNodes to neighbors (#247)

timoxley committed:
test(client, network): Fix tsconfig references.

timoxley committed:
test(protocol): Fix tsconfig references.

timoxley committed:
deps: Refresh package-locks.

timoxley committed:
deps: Refresh package-locks.

timoxley committed:
build(test-utils): Fix tsconfig references.

timoxley committed:
build(network): Split DEFAULT_MAX_NEIGHBOR_COUNT into separate file so don't have to pull in Tracker for browser builds.

timoxley committed:
build(client): Tweak webpack config.

timoxley committed:
test(broker): Fix up weird old rootDirs import paths.

timoxley committed:
style(network): Linting.

timoxley committed:
test(network): Enable jest-extended in karma tests.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  network: extract DisconnectionManager from Node (#251)
  refactor(network): rm dead code (#252)
  ci(deps): bump actions/checkout from 2.3.4 to 2.3.5 (#253)

timoxley committed:
test(client): Tidy browser test output.

timoxley committed:
release(client): client@6.0.0-alpha.12

timoxley committed:
Revert "Merge branch 'main' into NET-542-brubeck-client-browser"

This reverts commit b4105d6fce8fa5375d44f5d4de71be187bd977e2, reversing
changes made to c4548ebe7a59d38c91e767203007a0acacdcaa56.

timoxley committed:
Revert "Merge branch 'main' into NET-542-brubeck-client-browser"

This reverts commit bd9142106fe4932b450a0aeee97d2564d8be7297, reversing
changes made to 560ea878c91f3045abfa50c870a414fc52eb5352.

timoxley committed:
release(client): client@6.0.0-alpha.13

timoxley committed:
test(broker): Add temporary revert merge conflict fix.

test(broker): Add temporary revert merge conflict fix.

timoxley committed:
test(client): Fix up tsc arguments in test-types script.

timoxley committed:
test(client): Fix up paths in test-exports script.

timoxley committed:
ci(client): Try not stopping server before starting it.

timoxley committed:
ci(client): Try not using pkill to stop browser test server.

timoxley committed:
deps: Refresh package-locks.

timoxley committed:
test(client): Fix up rethrown error handling test behaviour.

timoxley committed:
build(client, network): Fix logger to understand numeric module.id.

timoxley committed:
release(client, network): client@6.0.0-alpha.14, network@30.0.2-alpha.2

timoxley committed:
fix(client): Don't require full prefixed stream name in getOrCreateStream.

timoxley committed:
release(client): client@6.0.0-alpha.15.

timoxley committed:
fix(client): Shim formatWithOptions to work in browser.

timoxley committed:
release(client): client@6.0.0-alpha.16.

timoxley committed:
fix(client): Ensure resendSubscribe does not error if no storage assigned.

timoxley committed:
release(client): client@6.0.0-alpha.17.

teogeb commented:
Good improvements 👍 

> #### Also Note
> * Some diff noise due to reverting some merges from main.

If this PR shouldn't revert the changes when we merge it to  main, should we exclude those reverts in this PR (that is, revert the reverts)?

timoxley committed:
build: Prevent endless tsc rebuilds in watch mode by setting test build configs to noEmit: true.

timoxley committed:
build: Convert tsconfig path references to ../relative paths suitable for hoisted deps from npm workspaces.

timoxley committed:
build: bootstrap monorepo using npm workspaces instead. Workspaces in npm 8 appear functional.

timoxley committed:
deps: Use correct types package for node-fetch 2.x.

timoxley committed:
build: Repurpose npm run bootstrap to only install top-level dependencies.

timoxley committed:
docs: Update readme for npm workspaces.

timoxley commented:
> If this PR shouldn't revert the changes when we merge it to main, should we exclude those reverts in this PR (that is, revert the reverts)?

Yep, we should revert the reverts, though I'll try just rebasing the revert commits out. Hopefully that won't generate any hairy conflicts.

timoxley committed:
build: Update boostrap-pkg & Dockerfile for npm workspaces.

timoxley committed:
ci: Update github actions for npm workspaces.

timoxley committed:
build: Improved bootstrap/clean scripts.

timoxley committed:
ci: Use npm@8 in CI.

timoxley committed:
ci: Use node 16 for broker tests.

timoxley committed:
build: Prevent versions output being printed multiple times.

timoxley committed:
build(client): Fix webpack lookup paths for npm workspaces.

timoxley committed:
build(cross-client-testing): Remove unneeded node/npm/node_modules handling.

timoxley committed:
build: Add missing test-utils tsconfig references.

timoxley committed:
test(client): Fix test-exports webpack lookup paths for npm workspaces.

timoxley committed:
test(client): Convert nightwatch.json to .js so can look up chromedriver bin path.

timoxley committed:
ci(network): Split out failing network browser build. Use node 16.

timoxley committed:
test(client): Fix test-browser-types tsconfig.

timoxley committed:
test(client): Fix client onError test.

timoxley committed:
Revert "Revert "Merge branch 'main' into NET-542-brubeck-client-browser""

This reverts commit 41cbd3ff4e361a41d11bbf7b08155ad099c4d9fd.

timoxley committed:
Revert "Revert "Merge branch 'main' into NET-542-brubeck-client-browser""

This reverts commit fb1bc83671bc1319a888527ff212ea18c258bc39.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main: (48 commits)
  style(client): fix eslint error
  Fix broken browser tests (#269)
  client [NET-571]: Using the Brubeck client without providing a private key (#268)
  ci: specify exact action versions for easy debugability (#266)
  ci(deps): bump actions/checkout from 2.3.5 to 2.4.0 (#271)
  docs(broker): fix mqtt example
  fix(broker): rely on default value for config.network.webrtcDisallowPrivateAddresses
  fix(broker): fix default value type in schema
  release: cli-tools 6.0.0-alpha.0
  fix: cli-tools work with alpha client
  test(network): fix describe
  Update README.md
  style(broker): break up log line
  ci(test-utils): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(protocol): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(network): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(client-dataunions): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(client-code): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(client-build): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(broker): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ...

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser: (50 commits)
  Revert "Revert "Merge branch 'main' into NET-542-brubeck-client-browser""
  Revert "Revert "Merge branch 'main' into NET-542-brubeck-client-browser""
  style(client): fix eslint error
  Fix broken browser tests (#269)
  client [NET-571]: Using the Brubeck client without providing a private key (#268)
  ci: specify exact action versions for easy debugability (#266)
  ci(deps): bump actions/checkout from 2.3.5 to 2.4.0 (#271)
  docs(broker): fix mqtt example
  fix(broker): rely on default value for config.network.webrtcDisallowPrivateAddresses
  fix(broker): fix default value type in schema
  release: cli-tools 6.0.0-alpha.0
  fix: cli-tools work with alpha client
  test(network): fix describe
  Update README.md
  style(broker): break up log line
  ci(test-utils): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(protocol): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(network): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(client-dataunions): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ci(client-code): replace styfle/cancel-workflow-action with GitHub Actions concurrency
  ...

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  ci: Use nick-invision/retry@2.5.1 instead of nick-invision/retry@v2.5.1.
  broker [NET-573]: MQTT plugin improvements (#272)

timoxley committed:
fix(broker): Fix merge issues.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  fix(broker): Fix merge issues.
  ci: Use nick-invision/retry@2.5.1 instead of nick-invision/retry@v2.5.1.
  broker [NET-573]: MQTT plugin improvements (#272)

timoxley committed:
fix(broker): Fix merge issues.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  fix(broker): Fix merge issues.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  ci: Use nick-invision/retry@2.5.1 instead of nick-invision/retry@v2.5.1.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  ci: Use nick-invision/retry@2.5.1 instead of nick-invision/retry@v2.5.1.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  ci: Use nick-invision/retry@v2 instead of nick-invision/retry@2.5.1.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  ci: Use nick-invision/retry@v2 instead of nick-invision/retry@2.5.1.

timoxley committed:
fix(broker): Fix merge issues.

timoxley committed:
build: Refresh package-lock.

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  broker [NET-581]: Remove legacyPublishHttp and legacyMqtt plugins (#277)
  fix(client): unsubscribe from node #276
  network: add support for jest-extended in browser tests (#273)
  fix: getMemberStats return values in README

timoxley committed:
Merge branch 'main' into NET-542-brubeck-client-browser

* main:
  ci(broker): Push broker image with 'debug' tag if docker test steps fail.
  build: Try improve Dockerfile npm speed with npm@8 --no-audit --prefer-offline, pruning in parallel, only pruning broker + deps.
  build: Split Docker build stage into multiple layers for convenience of caching.
  ci: Upload docker log artifacts on failure in broker & cross-client.
  ci(broker): Separate build+test & build+push jobs. Means build+test doesn't need to wait for other tests.
  build(broker): Run npm prune step on separate docker layer.
  build(broker): Ignore scripts during npm prune in Dockerfile.
  ci(broker): Build images for non-main branch.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  ci(broker): Push broker image with 'debug' tag if docker test steps fail.
  build: Try improve Dockerfile npm speed with npm@8 --no-audit --prefer-offline, pruning in parallel, only pruning broker + deps.
  build: Split Docker build stage into multiple layers for convenience of caching.
  ci: Upload docker log artifacts on failure in broker & cross-client.
  ci(broker): Separate build+test & build+push jobs. Means build+test doesn't need to wait for other tests.
  build(broker): Run npm prune step on separate docker layer.
  build(broker): Ignore scripts during npm prune in Dockerfile.
  ci(broker): Build images for non-main branch.
  broker [NET-581]: Remove legacyPublishHttp and legacyMqtt plugins (#277)
  fix(client): unsubscribe from node #276
  network: add support for jest-extended in browser tests (#273)
  fix: getMemberStats return values in README

timoxley committed:
test(client): Fix realtime GroupKeyPersistence test.

timoxley committed:
Merge branch 'NET-542-brubeck-client-browser' into npm-workspaces

* NET-542-brubeck-client-browser:
  test(client): Fix realtime GroupKeyPersistence test.

timoxley committed:
Merge pull request #259 from streamr-dev/npm-workspaces

npm workspaces

timoxley committed:
ci: Update cache-dependency paths to remove packages/*/package-lock.json

timoxley committed:
ci(network): Fix network-browser workflow.

timoxley committed:
ci: Use node 16 instead of 14 when only testing a single node version.

timoxley committed:
ci: Rename 'npm ci' steps to 'bootstrap-pkg'.

timoxley committed:
ci(network): Fix network workflow.

timoxley committed:
ci: Use --prefer-offline --no-audit for npm install/ci steps.

timoxley committed:
ci(network): Remove redundant browser test job.

timoxley committed:
test(client): Try get SubscriberResendsSequential to be more reliable.

timoxley committed:
test(client): Don't flag dev deps used in non-test files in test dir, e.g. test/utils.ts.

timoxley committed:
test(client): Don't flag dev deps used in non-test files in test dir, e.g. test/utils.ts.

timoxley committed:
test(client): Improve SubscriberResendsSequential.

timoxley committed:
refactor(client): Throw if no streamMessage supplied to waitForStorage.

teogeb commented on .github/workflows/broker.yml:
```diff
@@ -26,21 +26,19 @@ jobs:
     strategy:
       matrix:
-        node-version: [ 14.x ]
+        node-version: [16.x]
```
Many tests are now running on Node 16 instead of Node 14. Would it make drop the Node 14 support for the remaining test matrixes, too (benchmark tests in `client-build`, flakey and memory tests in `client-code`, lint and unit tests in `protocol`, and `test-utils`)

timoxley replied:
Yeah not a bad idea but perhaps in another PR. Should probably switch to 16 + 17 instead of 14 + 16

teogeb commented on packages/broker/src/plugins/legacyWebsocket/RequestHandler.ts:
```diff
@@ -65,27 +71,28 @@ export class RequestHandler {
+            case ControlMessage.TYPES.ResendRangeRequest:
                 return this.resend(connection, request as (ResendLastRequest|ResendFromRequest|ResendRangeRequest))
-            default:
+            default: {
```
Is this `{ }` wrapper needed?

timoxley replied:
Not required in this case, but putting a block around multi-line switch cases is a good habit as it's low cost but prevents unexpected weirdness triggered by putting declarations in the case body: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#block-scope_variables_within_switch_statements

Likely TS would pick up any issues, so this habit is probably legacy JS, but I prefer indenting with the brackets there anyway.

teogeb commented on packages/client/src/utils/Context.ts:
```diff
@@ -12,7 +12,7 @@ export class ContextError extends Error {
     constructor(context: Context, message: string = '', ...args: any[]) {
         // @ts-expect-error inspectOpts not in debug types
-        super(`${context.id}: ${formatWithOptions({ ...context.debug.inspectOpts, colors: false }, message, ...args)}`)
+        super(`${context?.id}: ${formatWithOptions({ ...(context?.debug?.inspectOpts || {}), colors: false }, message, ...args)}`)
```
Should the `context` parameter be optional if `?` operation is needed?

timoxley replied:
it's not actually optional, but at runtime it could possibly be undefined.

This is often invoked like: 

```tyoescript
class Task implements Context {
  id = instanceId(this)
  debug = debug(this.id)
  run() {
    // …
    throw new ContextError(this, 'an error happened during run')  
  }
}
```

which ideally would spit out errors something like: `"Error: Task:1 an error happened during run"`

but if this is invoked in a manner like so:

```typescript
const task = new Task()
setTimeout(task.run)
```

then at runtime, the useful part of the error will replaced with `Uncaught TypeError: Cannot read properties of undefined (reading 'id')`. 

With the above change errors will instead look like: `"Error: undefined: an error happened during run"`. Ideally this doesn't happen at all, but if it does then the least we can do is to ensure the original error message/stack isn't thrown away or replaced.

**Key points**: Never throw away error information. Ensure errors can't throw errors.

Related to this, it's possible we should enable this typescript-eslint rule in order to help catch cases like `setTimeout(task.run)` at compile time: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md

teogeb commented on packages/client/src/utils/ConvertBrowserStream.ts:
```diff
@@ -0,0 +1,36 @@
+import { PassThrough, Readable } from 'stream'
+
+export function ConvertBrowserStream(browserStream: ReadableStream | Readable) {
```
The name could describe what it is conversion target? Now bit unclear what kind of transformation is done. (Also as it is just a function, the name could be lowercased)

timoxley replied:
Will do.

timoxley replied:
Done

teogeb replied:
Is there a reason why this doesn't follow the usual naming convention that the first letter of a function name is lower-cased?

teogeb commented on packages/client/src/utils/log.ts:
```diff
@@ -41,6 +41,14 @@ if (typeof window === 'undefined') {
     })
 }
 
+// e.g. in browser
```
Could this fallback handling be part of the `export function formatWithOptions` implementation at line 73 so that we don't need to modify an external object? It a bit surprising side-effect of the `log.ts` import.

timoxley replied:
Good point. Changed.

teogeb commented on packages/client/src/BrubeckNode.ts:
```diff
@@ -32,7 +30,7 @@ export default class BrubeckNode implements Context {
     ) {
         this.options = options
-        this.id = instanceId(this, uid)
+        this.id = instanceId(this)
```
Currently there is a `id` parameter in options, and it if the value is not set, we generate a fallback value. Maybe we could use that same id in here so that there's only one ID to refer to a BrubeckNode? 

(Maybe a separate PR, if we want to implement something like that)

timoxley replied:
That's the node id, which is a bit different. 

This `id` is really internal and mostly there for logs & debugging. It follows the pattern used in the rest of the client codebase. It allows differentiating between instances in logs e.g.
```javascript
const thing1 = new Thing(opts)
const thing2 = new Thing(opts)
thing1.debug('test') // Thing:1 test
thing2.debug('test') // Thing:2 test
```
and also logging "namespaces" so you can see who generated an instance:
```javascript
const child1 = thing1.createChild()
const child2 = thing2.createChild()
child1.debug('test') // Thing:1 Child: 1 test
child2.debug('test') // Thing:2 Child: 2 test
```

teogeb commented on packages/client/src/Ethereum.ts:
```diff
@@ -66,7 +67,7 @@ export abstract class EthereumConfig {
 
 @scoped(Lifecycle.ContainerScoped)
-export default class StreamrEthereum {
+class StreamrEthereum {
```
Is this change of export style needed for browser compatibility? If so, why only this class was changed?

timoxley replied:
IIRC I think there was some circular dependency hell and this resolved it.

teogeb commented on packages/client/src/ResendSubscribe.ts:
```diff
@@ -57,7 +57,10 @@ export class ResendSubscription<T> extends Subscription<T> {
             yield* await this.getResent()
         } catch (err) {
-            await this.handleError(err)
+            if (err.code !== 'NO_STORAGE_NODES') {
```
Maybe worth mentioning this change of functionality in the PR description?

timoxley replied:
This fixes a v6 regression.

teogeb commented on packages/client/test/integration/GroupKeyPersistence.test.ts:
```diff
@@ -145,10 +145,10 @@ describeRepeats('Group Key Persistence', () => {
 
             published.push(...await publishTestMessages(3))
-            const received2 = await sub2.collect(3)
+            const received2 = await sub2.collect(4)
```
Could be worth commenting why we need to collect 4 messages instead of 3. Is the reason is that the internal group key response will be visible for the subscribers? Should it be?

timoxley replied:
I'd add a comment if I knew why 🤷  it changed at some point. Just matching network/broker behaviour.

teogeb replied:
But is the current behavior a bug? 

If I understood correctly, the use case we are testing is:
- subscribe with `subscriber1` 
- publish 1 test message
- expect `subscriber1` to receive the test message
- subscribe with `subscriber2`
- publish 3 more test messages
- expect `subscriber2` to receive 4 messages (`expect(received2).toEqual(published)`)

As the first message was published before `subscriber2` subscribed, should it receive the first test message (published at line 130). I'd assume it should it just receive 3 messages which were published after it subscriber.

Maybe the test would be more readable if we didn't alter the `const published` variable, but instead have variables like `published1` and `published2`.

teogeb commented on packages/network/src/helpers/logger/LoggerBrowser.ts:
```diff
@@ -0,0 +1,39 @@
+        super(options, destinationStream)
+    }
+
+    private static createName(module: NodeJS.Module, context?: string) {
```
The `createName` functions in `LoggerBrowser` and `LoggerNode` are almost identical. Would it be possible to move the method e.g. to `LoggerCommon`?

timoxley replied:
Done in [dd64b882](https://github.com/streamr-dev/network-monorepo/pull/244/commits/0a7622c6dde8527acada1980264611f9febfae41)

timoxley committed:
refactor(network): Move shared Logger#createName into LoggerCommon.

timoxley committed:
refactor(client): Remove util.formatWithOptions polyfill. Just use log.ts exported method.

timoxley committed:
refactor(client): ConvertBrowserStream -> WebStreamToNodeStream. Improve implementation, add test.

timoxley committed:
refactor(client): Tidy copy-package script.

timoxley committed:
ci(broker): Run docker build+test on ubuntu-latest.

Self-hosted seems to have an issue with streamr-docker-dev.

timoxley committed:
ci(network): Fix missing node matrix version in browser test.

teogeb commented on packages/client/src/utils/WebStreamToNodeStream.ts:
```diff
@@ -0,0 +1,71 @@
+    reader.closed.finally(() => { // eslint-disable-line promise/catch-or-return
+        toNodeStream.off('close', cleanup)
+    })
+    // toNodeStream.once('error', cleanup)
```
Remove if not needed?

teogeb commented on packages/client/src/Publisher.ts:
```diff
@@ -147,6 +147,9 @@ export default class BrubeckPublisher implements Context, Stoppable {
         count?: number
         messageMatchFn?: (msgTarget: StreamMessage, msgGot: StreamMessage) => boolean
     } = {}) {
+        if (!streamMessage || !streamMessage.spid) {
```
As `streamMessage: StreamMessage` is not an optional parameter, it should be caller's responsibility to ensure that the method is called with a valid value. The method should be able to rely on the contract it defined in the parameter list definition.

timoxley committed:
test(client): Revert SubscriberResendsSequential test changes.

timoxley committed:
deps(client): Clean out unused dependencies.

timoxley committed:
refactor(client): Remove examples.

Closes NET-555

timoxley committed:
deps: Refresh package-lock.json.

timoxley committed:
build(broker): Remove lerna-based prune from Dockerfile.

timoxley committed:
build(broker): Do not copy node_modules, dist, build, logs, etc when building docker image.

timoxley committed:
build: Remove temporary scripts to pull remote changes into monorepo.

timoxley committed:
test(client): Lint & fix bad syntax in tests.

timoxley committed:
test: Use --cache for linting. Add top level eslint script. Remove lerna-based scripts.

timoxley committed:
build: Remove lerna.

timoxley committed:
build: Lint monorepo before pushing.

timoxley committed:
fix(broker): Prevent temporary storage node unassignment on initial assignment.

timoxley committed:
build: Fix show-versions when npm prints warnings.

timoxley committed:
build: Rename script clean-root to boostrap-root.

timoxley committed:
Revert "build: Remove lerna."

This reverts commit a01f0ef66ef075e02a2d29e80392d4ecff5396ba.

timoxley committed:
test(network): Replace test() with it() as possible workaround for 'test is not defined' issue in network browser tests.

timoxley committed:
test(network): Fix karma browser tests.

timoxley committed:
Revert "test(network): Replace test() with it() as possible workaround for 'test is not defined' issue in network browser tests."

This reverts commit 84efd9364df32cab3b1d20b108ff6ebf2924ac30.

timoxley committed:
test(client): Fix up Session unit test.

timoxley committed:
ci(client): Build & use current monorepo brokers/trackers for client integration tests.

timoxley committed:
test(broker): Fix tests around delayed setSPIDKeys in storage.

timoxley committed:
ci(client): Try running client integration tests on self-hosted runner.

timoxley committed:
test(client): Fix up utils unit test.

timoxley committed:
ci(client): Test client exports in CI.

timoxley committed:
fix: Address all eslint errors and warnings.

timoxley committed:
ci(client): Build & use current monorepo brokers/trackers for cross-client tests.

timoxley committed:
ci(client): Run single iteration of dataunion tests.

timoxley committed:
deps: Lint package.json dependencies with manypkg.

timoxley committed:
build: Incorporate manypkg into versions script. Add fix script to run manypkg+eslint fix.

timoxley committed:
ci(client): Run client integration tests & cross client tests on ubuntu-latest.

timoxley committed:
deps(client): Update client deps & unify versions. Include yarn so manypkg fix works.

timoxley committed:
fix(client): Tidy tests.

timoxley committed:
test(client): Use different clients for publishing+subscribing in SubscriberResendsSequential.

Setup:
Publishes 5 messages, waits until they’re stored by polling /last
endpoint.  Stores list of published messages.

Test:
Ask for resend + subscribe
Publish a message. Add this to the list of published messages.
Wait until we get the resent messages + the new message.

After test:
Wait for last message to be stored before proceeding to next test. The
old messages plus the new message now become what we expect to get from
the resend.
Repeat test 4x

What (usually) happens:
First test succeeds
Second test fails
Other tests usually succeed, but occasionally fail.

I suspected this was related to the storage node being unassigned then
reassigned, but I fixed that and I still saw the issue.
Failure only ever occurs in the “after test” phase, when it’s waiting
for the last message to land in storage. Increasing the “wait for
storage” timeout doesn’t help. Curiously, this message will usually
appear in the resent data once the test starts.

I logged when messages were hitting storage and the missing message
wasn’t being stored until after the “wait for last” step times out. If I
increase the timeout, the message takes longer to store.

What’s happening is that the message isn’t being propagated from the
node to storage until the next test starts.

When the client finishes the resend, and sees it has no further
subscriptions, it unsubscribes the node from the stream before the
message was propagated. In the test, the publisher node is the
subscriber so the client is immediately passed the realtime message.

The test is "fixed" by making the publisher and subscriber
different nodes.

Underlying problem is that the node unsubscribes before it propagates
messages.

timoxley committed:
build: Fix eslint script to fail if eslint fails.

timoxley committed:
ci: Share broker docker image in cross-client tests.

timoxley committed:
build: Fix npm run clean-dist.

timoxley committed:
build: Fix npm run clean-dist.

timoxley committed:
ci: Fix share broker docker image in cross-client tests.

timoxley committed:
ci(cross-client): No need to bootstrap before docker build.

timoxley committed:
build(client): Don't transpile esm index.

timoxley committed:
test(client): Get MultipleClients test to be reliable.

timoxley committed:
build(client): Tidy up config files.

timoxley committed:
test(client): Relax until condition.

timoxley committed:
test(client): Try different tactic to get MultipleClients test more reliable.

timoxley committed:
test(client): Try different tactic to get MultipleClients test more reliable.

timoxley committed:
build(client): Tidy up package.json.

timoxley committed:
test(client): Try running resend tests in series.

timoxley committed:
test(client): Skip late subscriber tests.
